### PR TITLE
Support negative timestamps for the query engine

### LINKED ISSUE #7194: can't query for points before epoch 0
### Bug report

__System info:__ [Include InfluxDB version, operating system name, and other relevant details]
Ubuntu 14.01, InfluxDB 0.13 from package

__Steps to reproduce:__

Add some data, including data before epoch 0
```
Connected to http://localhost:8086 version 0.13.0
InfluxDB shell version: 0.13.0
> use mydb
Using database mydb
> insert e0 value=1 -100
> insert e0 value=2 100
> insert e0 value=3
```

As expected, epoch 0 is the lower bound if none is provided:
```
> select * from e0
name: e0
--------
time			value
100			2
1471979518382460726	3

> 
```

However, the lower bound cannot be changed to anything less than epoch 0:

```
> select * from e0 where time > '1968-01-01'
ERR: time 1968-01-01T00:00:00Z underflows time literal
> select * from e0 where time > '1968-01-01T00:00:00Z'
ERR: time 1968-01-01T00:00:00Z underflows time literal
> select * from e0 where time > '1970-01-01T00:00:00Z'
name: e0
--------
time	value
100	2

> select * from e0 where time > '1969-01-01T00:00:00Z'
ERR: time 1969-01-01T00:00:00Z underflows time literal
> select * from e0 where time > '1969-12-31T23:59:59Z'
ERR: time 1969-12-31T23:59:59Z underflows time literal
> select * from e0 where time > now() - 4000w
ERR: time 1939-12-26T19:12:10Z underflows time literal
```

__Expected behavior:__ [What you expected to happen]

If negative timestamps are valid for writes, they need to be valid for reads.

### PR DESCRIPTION
**Pull request by jsternberg**
Fixes #7194.

### CODE REVIEW

**Comment by jsternberg** [comment]
Need to fix a bunch of tests, but this seems to work in local testing.

File: influxql/ast.go [hunk 1]
```diff
@@ -3630,7 +3630,7 @@ func TimeRange(expr Expr) (min, max time.Time, err error) {
 }
 
 // TimeRangeAsEpochNano returns the minimum and maximum times, as epoch nano, specified by
-// and expression. If there is no lower bound, the start of the epoch is returned
+// and expression. If there is no lower bound, the minimum time is returned
```
Comment (joelegasse) [inline influxql/ast.go]: You updated the comment here, but line 3642 still sends the min to `time.Unix(0,0)` when the min time is unset.

File: influxql/iterator.go [hunk 2]
```diff
@@ -19,7 +19,7 @@ var ErrUnknownCall = errors.New("unknown call")
 
 const (
 	// MinTime is used as the minimum time value when computing an unbounded range.
-	MinTime = int64(0)
```
Comment (joelegasse) [inline influxql/iterator.go]: Why is the max time copied exactly and the min time decreased by one? Can you add comment to clarify why this is being done?

File: influxql/ast.go (hunk 1 above)
Comment (jsternberg) [inline influxql/ast.go]: Good catch. I went through a bunch of different revisions while doing this and I might have lost some things like this.

File: influxql/point.go [hunk 3]
```diff
@@ -9,10 +9,11 @@ import (
 
 	"github.com/gogo/protobuf/proto"
 	internal "github.com/influxdata/influxdb/influxql/internal"
+	"github.com/influxdata/influxdb/models"
 )
 
-// ZeroTime is the Unix nanosecond timestamp for time.Time{}.
-const ZeroTime = int64(-6795364578871345152)
+// ZeroTime is the Unix nanosecond timestamp for no time.
+const ZeroTime = models.MinNanoTime - 2
```
Comment (joelegasse) [inline influxql/point.go]: Please add a comment for the arbitrary subtraction of two.

File: models/time.go [hunk 4]
```diff
@@ -12,9 +12,9 @@ import (
 const (
 	// MinNanoTime is the minumum time that can be represented.
 	//
-	// 1677-09-21 00:12:43.145224192 +0000 UTC
+	// 1677-09-21 00:12:43.145224194 +0000 UTC
 	//
-	MinNanoTime = int64(math.MinInt64)
+	MinNanoTime = int64(math.MinInt64) + 2
```
Comment (joelegasse) [inline models/time.go]: Please add something explaining the addition of two here, as well.

**Comment by jsternberg** [comment]
@joelegasse I tried my best to add documentation for why the arbitrary values for minimum and maximum times.

**Comment by joelegasse** [comment]
Thanks for adding the comments. LGTM

**Comment by benbjohnson** [comment]
👍

**Pull request merged**